libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts

```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
+            const propName = this.stripProperty(name);
+            let entityName: string;
+            let isOp = false;
+            let isProperty = false;
+            if (operations.includes(name)) {
+                op = name;
+                isOp = true;
+            } else if (properties.includes(propName)) {
+                property = propName;
+                isProperty = true;
+            } else {
+                entityName = name;
+            }
 
-                    if (propertyName.end > info.end) {
-                        info.end = propertyName.end;
-                    }
+            entities.forEach((entity, index) => {
+                const instance = instances[index];
+                let root = rootInstance;
+                if (!root) {
+                    // Keep the root entity name and position to help with overlap.
+                    root = cloneDeep(instance);
+                    root.type = `${name}${index}`;
+                }
 
-                    // Expand entity to include possible property names
-                    for (const property in propertyName.value) {
-                        const newInfo: Partial<EntityInfo> = Object.assign({}, info);
-                        newInfo.property = property;
-                        infos.push(newInfo);
-                    }
-                } else {
-                    if (op && name == this.propertyNameKey) {
-                        for (const property in val) {
-                            const newInfo: Partial<EntityInfo> = Object.assign({}, info);
-                            newInfo.property = property;
-                            infos.push(newInfo);
+                if (entityName) {
+                    this.expandEntity(
+                        entityName,
+                        entity as Record<string, unknown[]>,
+                        instance,
+                        root,
+                        op,
+                        property,
+                        turn,
+                        text,
+                        entityToInfo
+                    );
+                } else if (typeof entity === 'object' && entity !== null) {
+                    if (isEmpty(entity)) {
+                        if (isOp) {
+                            // Handle operator with no children.
+                            this.expandEntity(op, null, instance, root, op, property, turn, text, entityToInfo);
+                        } else if (isProperty) {
+                            // Handle property with no children.
+                            this.expandEntity(property, null, instance, root, op, property, turn, text, entityToInfo);
                         }
                     } else {
-                        infos.push(info);
+                        this.expandEntityObject(
+                            entity as Record<string, unknown[]>,
```

Comment:
Whenever I'm forced to use `as T` style casting, I like to leave a comment explaining why.